compiler/src/dotty/tools/io/FileWriters.scala

```diff
@@ -25,13 +24,107 @@ import java.util.zip.CRC32
 import java.util.zip.Deflater
 import java.util.zip.ZipEntry
 import java.util.zip.ZipOutputStream
-import scala.language.unsafeNulls
+import scala.collection.mutable
+
+import dotty.tools.dotc.core.Contexts, Contexts.Context
+import dotty.tools.dotc.core.Decorators.em
+
+import dotty.tools.dotc.util.{SourcePosition, NoSourcePosition}
+
+import dotty.tools.dotc.reporting.Message
+import dotty.tools.dotc.report
+
+import dotty.tools.backend.jvm.PostProcessorFrontendAccess.BackendReporting
+import scala.annotation.constructorOnly
 
 /** Copied from `dotty.tools.backend.jvm.ClassfileWriters` but no `PostProcessorFrontendAccess` needed */
 object FileWriters {
   type InternalName = String
   type NullableFile =  AbstractFile | Null
 
+  inline def ctx(using ReadOnlyContext): ReadOnlyContext = summon[ReadOnlyContext]
+
+  sealed trait DelayedReporting {
+    def hasErrors: Boolean
+    def error(message: Context ?=> Message, position: SourcePosition): Unit
+    def warning(message: Context ?=> Message, position: SourcePosition): Unit
+    def log(message: String): Unit
+
+    def error(message: Context ?=> Message): Unit = error(message, NoSourcePosition)
+    def warning(message: Context ?=> Message): Unit = warning(message, NoSourcePosition)
+  }
+
+  final class EagerDelayedReporting(using captured: Context) extends DelayedReporting:
+    private var _hasErrors = false
+
+    def hasErrors: Boolean = _hasErrors
+
+    def error(message: Context ?=> Message, position: SourcePosition): Unit =
+      report.error(message, position)
+      _hasErrors = true
+
+    def warning(message: Context ?=> Message, position: SourcePosition): Unit =
+      report.warning(message, position)
+
+    def log(message: String): Unit = report.echo(message)
+
+  final class BufferingDelayedReporting extends DelayedReporting {
+    // We optimise access to the buffered reports for the common case - that there are no warning/errors to report
+    // We could use a listBuffer etc - but that would be extra allocation in the common case
+    // Note - all access is externally synchronized, as this allow the reports to be generated in on thread and
+    // consumed in another
+    private var bufferedReports = List.empty[Report]
+    private var _hasErrors = false
+    enum Report(val relay: Context ?=> BackendReporting => Unit):
+      case Error(message: Context => Message, position: SourcePosition) extends Report(ctx ?=> _.error(message(ctx), position))
+      case Warning(message: Context => Message, position: SourcePosition) extends Report(ctx ?=> _.warning(message(ctx), position))
+      case Log(message: String) extends Report(_.log(message))
+
+    def hasErrors: Boolean = synchronized:
+      _hasErrors
```

Comment:
IIRC, `hasErrors` gets called quite often in the compiler. Is this one of the possible targets of all these calls? If yes, the `synchronized` overhead might be significant.

Consider using an `AtomicBoolean` instead.

(In fact, it looks like the usage pattern `bufferedReports` is a textbook example of an `AtomicReference` with `compareAndSet`s.)

Reply:
this will get called one time in the `writeSigFilesAsync`

Reply:
updating to use atomic references